Add Testing Mode and ignoreMotion function

veeceeoh opened the pull request:
As explained in [my post on the SmartThings forum](https://community.smartthings.com/t/original-aqara-xiaomi-zigbee-sensors-contact-temp-motion-button-outlet-leak-etc/113253/705), I've added a second countdown timer so that all motion detected messages are ignored for a period of 5 seconds less than the motion reset timer. The state of ignore or not is tracked by a "global" state object, `state.ignoreMotion`.

This avoids issues with automations being activated every 5-6 seconds when the sensor is put into "testing mode."

I have also changed the default motion reset preference setting to 61 seconds, so that the motion reset counter can be extended another 60 seconds if the sensor detects movement right after its hardware "blind" period finishes.

I removed `state.battery = 0` from installed() and configure() because it's not used at all, and added `state.ignoreMotion == false` to those calls as well as `updated()` to ensure the DTH pairs the sensor or saves preference settings with motion detected messages NOT being ignored.

alecm commented:
Sorry, @veeceeoh.  I appreciate what you're trying to do, but I'm not convinced that there's any real need for the second countdown timer - and it adds even more complexity to the code.  In my case as a somewhat advanced end user I can't even figure out how to test for this. I do *not* believe that the test mode should be circumvented. Happy to work on a solution that notifies the end user that they are in test mode.

veeceeoh commented:
I think I'm just not explaining it well enough.

What I'm proposing with the code change does not circumvent the test mode at all.

Regardless of whether this change is made or not, while the sensor is in test mode, with the default setting of 60 seconds, in the mobile app you won't see immediate visual feedback that the sensor has stopped detecting motion.

Because of this users should change the motion reset to a much lower number of seconds than the 60-second default if they want to check if the installation location will work well. I agree it would be great to notify users that the sensor is in test mode, but it would be very difficult to implement. I think in this case it's better to let people know with documentation or extra explanation in the preferences page.

However, this change I'm proposing does not address this at all. The only thing it changes is to make sure the hub doesn't pass on motion detected events every 6 seconds when the motion reset preference is set higher than that or at the default of 60 seconds. The intention of the change is to prevent problems for users with an already installed sensor that has accidentally/unintentionally been put into test mode.

So this code ensures that the number of seconds set in the motion reset preference setting is honored by the DTH. The second countdown timer at 5 seconds less than the motion reset setting is just my solution to make sure it doesn't interfere with the test mode or the normal 60-second hardware reset.

Testing this is pretty simple: Just make sure the sensor works the way you'd expect, for example:

A. When in normal mode and the motion reset preference is set to 61 seconds, after it detects motion, and you're still in the room and moving 60 seconds later, does the app continue to show motion detected for another 60 seconds with no change to inactive?

B. If you put the sensor into test mode, and then change the motion reset time to 6 or 7 seconds, do you see the app tile go from motion detected to inactive a few seconds later when there's no actual motion going on?

If yes, then it's working. 

Again, to sum up, this code does not interfere at all with the test mode. But to make use of the test mode properly, users should be informed that it's best to change the motion reset preference setting to a small number like 6 or 7 seconds. And after testing is complete, change it back to a higher number or the default of 61 seconds.

alecm commented:
I think I get it now @veeceeoh - I've had my coffee :-).  I'd like to test over weekend (and suggest others test too) before you push. Preference page text might need to be looked at.

veeceeoh commented:
I'm doing testing now, and can immediately see that I need to rework my changes a bit. I will post again when that's ready to try.

veeceeoh commented:
I have just updated the code - it's working awesome now and with inspiration from @alecm there's a new setting: Testing Mode! (see below change list for screenshots.)

**Changes**
• Added a new Testing Mode preference setting toggle, with a VERY LONG and hopefully HELPFUL explanation
• The Testing Mode overrides the `motionreset` value to be 7 seconds, which works best with the 6 second reset of the hardware Test Mode
• The `ignoreMotion` feature is completely disabled while Testing Mode is active
• While Testing Mode is active, the "Last Motion: date" text in the main tile is replaced with ">>>>>  Testing Mode is ACTIVE  <<<<<"
• The Testing Mode is automatically deactivated 60 minutes after it's turned on
• The user can manually turn off Testing Mode using the preference toggle a second time (which is why it's called a "toggle")
• Like the battery replaced date toggle, when the save button is pressed, the toggle is reset. I tried to find a way to have the code change it after 60 minutes, but that's just not possible in a DTH
• When preferences are saved, motion detected is reset to "no motion" and the 'ignoreMotion' flag is also reset, just to make sure any changes to `motion reset` or turning Testing Mode on/off can take immediate effect.

**Fixes**
• The `ignoreMotion` implementation is now working correctly
• If Testing Mode is off, for any `motion reset` value of 11 or more seconds, motion detected messages are ignored for 6 seconds less than the `motion reset` value. So for example, if `motion reset` is set to the default of 61 seconds, motion detected messages are ignored for 55 seconds. This prevents a flood of motion detected events from happening when the user is finished testing the location of their sensor, or when the reset button is unintentionally pressed on an already installed sensor. 
• If `motion reset` is set to 10 or fewer seconds, then `ignoreMotion` is not used, in case the user has decided to use that setting instead of the new Testing Mode while the sensor is in hardware Test Mode.

**Screenshots**

![screen 1](https://user-images.githubusercontent.com/33558908/37561613-2953d836-2a10-11e8-8a09-194ce223b79e.PNG)

![screen 2](https://user-images.githubusercontent.com/33558908/37561617-32af72b4-2a10-11e8-8f73-b3ca957775cc.PNG)

![screen 3](https://user-images.githubusercontent.com/33558908/37561620-352fc944-2a10-11e8-8730-8d97def10eb7.PNG)

alecm commented:
wow - you've been busy!!!   Sorry - haven't had a chance to test this - my main smartthings motion sensor in the bathroom died (much to my surprise - reported temp but motion sensor totally gone).  Anyhow had to use my testing xiaomi to replace.  Will key another one in. 
That said - yesterday I was exploring something which I wasn't able to get too far on before I had to leave the house. There is a cluster which as far as I can tell is always and only sent when the reset button is pressed  - cluster 0000, attrib 0005 which gives the model.  
I've only tested after pairing - but I have a hunch it would also be shown at initial pairing.
How about using this particular event as an automatic test mode indicator with a 2 hour timer?  

I can throw this into  new suggestion. 

Alec 
------
d5b99a48-1ec2-4624-b142-435eac76f60a  2:09:20 PM: debug Bathroom Door Motion Sensor III reported: cluster: 0000, attrId: 0005, model:lumi.sensor_motion.aq2
d5b99a48-1ec2-4624-b142-435eac76f60a  2:09:20 PM: debug Bathroom Door Motion Sensor III parsing: read attr - raw: 9CD401000034050042166C756D692E73656E736F725F6D6F74696F6E2E617132, dni: 9CD4, endpoint: 01, cluster: 0000, size: 34, attrId: 0005, encoding: 42, value: 6C756D692E73656E736F725F6D6F74696F6E2E617132

veeceeoh commented:
@alecm - I have figured out a 100% reliable way to determine whether the sensor is in hardware test mode! I can't believe I didn't think of it before my in-the-shower-half-asleep "Ah ha!" moment this morning.

All the code needs to do is track the time difference between the last motion detected message and the current one. If the time is less than 60 seconds (or better yet let's say less than 10 seconds), then the sensor is in hardware test mode, and whatever actions that need to be taken are based on that.

Although it's possible someone may pair a sensor and then put it in a dark corner, it's far more likely they're handling it when it's paired, so this method should help to inform them, either through a new UI tile, or another line in the main tile secondary text (that normally displays the last motion date/ time).

alecm commented:
Sorry been out of loop a bit, @veeceeoh.  Agree w/ seldom putting in a dark corner - I think I used to do it to avoid clutter before identifying battery reporting.  

Good idea on the less than 60 seconds putting into test -  but - what would you propose to do to determine it's over?   (particularly if, say, they had had a 'fast motion'  an hour into testing)?   You call it something like "Probably in Hardware Testing"  or something like that (that's awkard) and count down two hours, which is the documented test time.

The other thing is they could press the button again before testing mode was over - this *is* easy to identify with the cluster I discussed earlier - it does get reported if button pressed during testing mode.  What I can't confirm is whether this resets the clock or not (so another two hours from when you pressed rather than from the original test mode start). I haven't tested.  Am going to do now - I have a sensor at hand.
